### PR TITLE
Fix: Resend magic code option behaves inconsistently

### PR DESCRIPTION
**Pull request by Pujan92**
### Details
Changes applied to the magic code screen which shows the resend magic code link all the time and success/error messages will be shown below it.

### Fixed Issues

$ https://github.com/Expensify/App/issues/17864
PROPOSAL: https://github.com/Expensify/App/issues/17864#issuecomment-1520081863

### Tests

1. Enter email/phone number on the login screen and continue
2. On magic code screen Verify the link `Didn't receive a magic code?` remains available all the time
3. Verify the below scenarios(Success/Error respective message shown below the link)
     a. Error on invalid magic code
     b. Success message `Magic code sent` on click of `Didn't receive a magic code?` link
     c. Error on maximum magic code sent attempts

- [X] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

1. Enter email/phone number on the login screen and continue
2. On magic code screen Verify the link `Didn't receive a magic code?` remains available all the time
3. Verify the below scenarios(Success/Error respective message shown below the link)
     a. Error on invalid magic code
     b. Success message `Magic code sent` on click of `Didn't receive a magic code?` link
     c. Error on maximum magic code sent attempts

- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/14358475/237039839-3bd4e838-d6b3-4ba3-90a4-9fcea93e2627.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

Screenshot_1683624125

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/14358475/237043025-74c5f042-58a7-4328-8b62-0f2b49629041.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/14358475/237041911-85cfe271-d47b-4eff-8dd0-61f2f13355b1.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/14358475/237041971-f6554dda-7a4c-4ce2-9504-4d251f943806.mp4

</details>

<details>
<summary>Android</summary>

Screenshot_1683623485

</details>

### CODE REVIEW

**Comment by Pujan92** [comment]
> @Pujan92 Tests are incomplete. Please fix that.

@parasharrajat are you referring Offline tests here?

**Comment by Pujan92** [comment]
@parasharrajat I updated the tests, is it makes more sense now?

> 1. Enter email/phone number on the login screen and continue
> 2. On magic code screen Verify the link `Didn't receive a magic code?` remains available all the time
> 3. Verify the below scenarios(Success/Error respective message shown below the link)
>    a. Error on invalid magic code
>    b. Success message `Magic code sent` on click of `Didn't receive a magic code?` link
>    c. Error on maximum magic code sent attempts

**Comment by Pujan92** [comment]
> Please add the clicking the resend button should show the message below it.

I think it is covered here
> b. Success message Magic code sent on click of `Didn't receive a magic code?` link

**Comment by shawnborton** [comment]
For this message here:
image


I think the text should use textSupporting and not red. We discussed that only form errors would get the full red text treatment (cc @JmillsExpensify @trjExpensify )

Also, for the text that says the link was sent, we should use the same dot indicator message style but the dot would be green and the message would be in textSupporting. 

> I noticed that the UI is moving up and down when the Magic Code sent message hides/unhides. 

Is there a way to make it so that the form is anchored at the top and it just makes the form either get longer/shorter? Cc @grgia - not sure how we are laying out the sign up form but maybe that's not possible given the vertical alignment?

**Comment by danieldoglas** [comment]
@parasharrajat can bring that discussion on slack in the open-source channel, please?

**Comment by grgia** [comment]
@shawnborton @parasharrajat, yes we should be able to anchor the form to the top. I believe right now we're using some outdated flex styling from before when this page required keyboard avoiding. We can definitely fix that.

**Comment by strepanier03** [comment]
@parasharrajat and @grgia - little bump to keep this from getting stale. Thank you!

**Comment by Pujan92** [comment]
> > Also, for the text that says the link was sent, we should use the same dot indicator message style but the dot would be green and the message would be in textSupporting.
> 
> @Pujan92 Can you do this?

Ok.

> > we should be able to anchor the form to the top. I believe right now we're using some outdated flex styling from before when this page required keyboard avoiding. We can definitely fix that.
> 
> @Pujan92 We are saying that we fix the starting gap on the Form so that when content changes (error messages or `link sent` messages hides or shows), the form's content does not move in the up direction. Only the bottom part is pushed down as the form content length increases.

@parasharrajat To confirm we need to take out the top empty View which is added here. That helps to not move content to upwards when the content gets changed.
https://github.com/Expensify/App/blob/1ab36df9fa03a3e85cb68c64de71eca3d559a3dd/src/pages/signin/SignInPageLayout/SignInPageContent.js#L46-L47

**Comment by Pujan92** [comment]
@parasharrajat @grgia @shawnborton Currently top empty view height is dynamically getting changes based on the below content with the flex properties. But now we are not going to use the same then what fixed top margin do we need to set for making it a bit down?

Screenshot 2023-05-24 at 11 57 51 AM

Screenshot 2023-05-24 at 11 57 43 AM

My concern is if we set a high fixed top margin to make it center and then when we display the error the content may go down and the scrollbar gets added.

**Comment by Pujan92** [comment]
@parasharrajat for final clarification I need to do these 2 things
1) Remove empty top View and provide fixed margin

**Comment by Pujan92** [comment]
> > Remove empty top View and provide a fixed margin (for eg. mt6)
> 
> I would say leave it if this is pushing the content too much down. But if it is done, let me test this first.

For testing, I have given a top margin of 50px and seems good, or for web/desktop we can provide some more margin.

Screencasts

https://github.com/Expensify/App/assets/14358475/065ea349-15d4-46fa-9a67-73ac318c942e

https://github.com/Expensify/App/assets/14358475/8270fabe-57c1-45bb-9bf5-573b96656b5c

**Comment by Pujan92** [comment]
@parasharrajat Ok, other point of using the DotIndicator has already been committed. Let me know if anything more I need to update in the PR.

**Comment by danieldoglas** [comment]
@Pujan92 let's just fix this merge conflict we have now, and @parasharrajat let's do a final review. I would be awsome to merge this today.

**Comment by Pujan92** [comment]
> @Pujan92 let's just fix this merge conflict we have now, and @parasharrajat let's do a final review. I would be awsome to merge this today.

@parasharrajat @danieldoglas Merge conflict is resolved.

**Comment by danieldoglas** [comment]
@parasharrajat please review this PR when you have the time!

**Comment by strepanier03** [comment]
@parasharrajat - This is waiting for your review. Will you have time to review it tomorrow so we can keep this moving forward?

**Comment by Pujan92** [comment]
> BUG: Clicking the resend link does not show the resent message when a wrong code is entered.

@parasharrajat Fixed the issue in the latest commit by clearing the existing magic code on resend function which will clear the field before making the api call. 

https://github.com/Expensify/App/assets/14358475/27be0077-c00e-42a9-9db3-6f6f88e0bfdd

**Comment by Pujan92** [comment]
Bump @parasharrajat

**Comment by strepanier03** [comment]
@parasharrajat - Will you be able to review this soon? If you're feeling swamped I can help find another team member, just let me know what I can do to help out and I'll try my best.

**Comment by strepanier03** [comment]
@Pujan92 - Little bump [here](https://github.com/Expensify/App/pull/18632#issuecomment-1577687764) to keep us moving forward

**Comment by Pujan92** [comment]
> I am trying to understand the new code which sounds like a workaround. Will post an update in some time. We will close this issue today.

I am not sure why this sounds like a workaround to you, isn't it good to clean existing magic code and message as soon as user presses the resend magic code link?

**Comment by danieldoglas** [comment]
@parasharrajat let's merge this today 🙏

**Comment by Pujan92** [comment]
> @Pujan92 Any take on this [#18632 (comment)](https://github.com/Expensify/App/pull/18632#issuecomment-1586616446)?

I will check in an hour and update you on it.

**Comment by Pujan92** [comment]
> Ok. Thanks. @Pujan92 can you confirm if this is present on staging or PROD?

@parasharrajat Currently, in staging/prod we are hiding the link on the press so it won't be tested there, but if in the main branch I show the link all time as per this PR change then it can be reproduced.

[aaad.webm](https://github.com/Expensify/App/assets/14358475/1d3acb50-17c1-43be-8ceb-72418054f787)

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js
Comment (Pujan92) [reply src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js:118]: Ok, basically the issue persists when we clear the `validateCode` the hook gets called and reset the `linkSent` state variable which we set to `true` in our resend function. To solve this I only found this way and did this. I will add a comment for it.

**Comment by Pujan92** [comment]
> @Pujan92 Bump.

Updated. Plz review.

File: src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js
Comment (Pujan92) [reply src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js]: We can add, I thought it will again create the same issue of linkSent but in particular this part it is relying on `prevIsVisible && !props.isVisible`. I will add it.

**Review by parasharrajat — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

cc: @danieldoglas 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Review by danieldoglas — APPROVED** [review]
LGTM

**Pull request merged**
